### ISSUE TITLE
Emulator

### ISSUE THREAD

**Issue by shandshellin** [DustinBrett/daedalOS#379]
Is there any way to add additional emulators for roms?

**Comment by glitch128** [comment]
which emulator do you mean?

**Comment by shandshellin** [comment]
the bundle as in what retroarch houses. There is already dos emulation in daedalos which is fantastic but i would love to run snes, mame, etc. Trying to get that all in web instance. Do game controllers work via the browser? Could i map a directory via docker so i could access me dos games from a share?

**Comment by glitch128** [comment]
> i would love to run snes, mame, etc.

There is emulator.js built into the OS

> Do game controllers work via the browser?

I have not tried that yet, but emulator.js does have gamepad support by itself

> Could i map a directory via docker so i could access me dos games from a share?

I have no idea

**Comment by DustinBrett** [comment]
Indeed Emulator.js already has support (.32x, .a26, .a52, .a78, .gb, .gba, .gbc, .gen, .gg, .j64, .jag, .lnx, .n64, .nds, .nes, .ngc, .ngp, .pce, .sfc, .smc, .smd, .sms, .v64, .vb, .vboy, .ws, .wsc, .z64) for lots of roms built into the OS.

And for mapping of folders, there is the ability to share external folders via the https://developer.mozilla.org/en-US/docs/Web/API/File_System_API, so if the browser has access to folders on the host machine, they can be shared inside daedalOS.

**Comment by glitch128** [comment]
Maybe you could add a feature to connect to one drive or Google Drive in the files app

**Comment by glitch128** [comment]
Maybe you could add a feature to connect to one drive or Google Drive in the files app

**Comment by DustinBrett** [comment]
> Maybe you could add a feature to connect to one drive or Google Drive in the files app

Ya that is the plan eventually, to integrate One Drive, and eventually others like Google/Dropbox. I haven't had a lot of personal use for the feature so I have not yet added it, and also I have avoided features which will require forked users to have their own API keys to make them work.

In a perfect world I would not need an API key for these features and I could just have the users give their creds when needed, but because that isn't how it works I would need to have my own API key and be registered with MS/Google.

**Comment by shandshellin** [comment]
you are correct, the roms all work except my mame roms. WIsh i could get unreal tournament on this!

**Comment by DustinBrett** [comment]
> you are correct, the roms all work except my mame roms. WIsh i could get unreal tournament on this!

Ya I will probably add MAME one day, but the roms I had examples of were multi file and for now I decided to focus on roms with single files.

As for UT, I do want to get it added one day, but the versions I have seen online don't work too well and I don't yet know the process to port it myself.

A demo is here of what they had many years ago: https://icculus.org/ut99-emscripten/flyby/wasm/

**Comment by shandshellin** [comment]
That demo is what makes me hopeful. My mame roms are all over the place as well. Any chance of maybe some open office?

**Comment by DustinBrett** [comment]
> That demo is what makes me hopeful. My mame roms are all over the place as well. Any chance of maybe some open office?

Ya Open Office is another possibility, the main reason I haven't done that one yet is the port I have seen is a somewhat large file and also it's not setup well to be interactive with daedalOS.

Here is the demo I have seen although it's LibreOffice instead of Open Office: https://wasm-test.libreoffice.org/
